core/bentley/src/BentleyError.ts

```diff
@@ -315,20 +378,37 @@ export type GetMetaDataFunction = () => object | undefined;
  */
 export type LoggingMetaData = GetMetaDataFunction | object | undefined;
 
-function isObject(obj: unknown): obj is { [key: string]: unknown } {
-  return typeof obj === "object" && obj !== null;
-}
 
 interface ErrorProps {
   message: string;
   stack?: string;
   metadata?: object;
 }
 
-/** Base exception class for iTwin.js exceptions.
+/**
+ * An `ITwinError.Error` instance that also supplies an `errorNumber`.
+ * @note this interface exists *only* for legacy errors derived from `BentleyError`. The concept of "error number" is
+ * problematic since it is impossible to enforce across the iTwin.js library, let alone across applications. New code should
+ * use `ITwinError` and identify errors with strings instead.
+ * @beta */
+export interface LegacyITwinErrorWithNumber extends ITwinError.Error {
+  /** a number to identify the error. */
+  readonly errorNumber: number;
+
+  /** Logging metadata
+   * @note exceptions should *not* include logging data. Logging should be done where exceptions are caught. This member exists
+   * only for backwards compatibility.
+   */
+  loggingMetadata?: object;
+}
+
+/**
+ * Base exception class for legacy iTwin.js errors.
+ * For backwards compatibility only. Do not create new subclasses of BentleyError. Instead use [[ITwinError]].
  * @public
  */
-export class BentleyError extends Error {
+export class BentleyError extends Error { // note: this class implements LegacyITwinErrorWithNumber but can't be declared as such because that interface is @beta.
```

Comment:
wondering if we should promote LegacyITwinErrorWithNumber to public but also marked as @deprecated to be more explicit

Reply:
This new interface shouldn't be deprecated - it will have to live forever since we can't get rid of BentleyError.

Reply:
@aruniverse satisfied?

Reply:
I would expect that we eventually deprecate BentleyError in favor of ITwinError. However, ITwinError is being introduced as beta, so BentleyError shouldn't be deprecated in this PR.

Your original suggestion of introducing and immediately deprecating LegacyITwinErrorWithNumber seems weird to me. Why introduce a brand-new public API that nobody is supposed to use?

Reply:
I'm not sure how we can make BentleyError deprecated. The problem is that it should be an interface not a class. 

I suppose we could start by making its constructor deprecated and adding a `throwError` method to it. The main thing we need to do is stop people from creating subclasses and doing `instanceOf` tests for it, and I don't know how to do that.

Reply:
Regardless of whether/how we do that, I don't think such changes belong in this PR.